### PR TITLE
Added upsertMany function

### PR DESCRIPTION
**Pull request by dsurrao**
Implemented upsertMany() based on this thread: https://community.openfn.org/t/upsertmany-for-the-mysql-adaptor/111

### CODE REVIEW

**Comment by taylordowns2000** [comment]
This looks great. @stuartc , @lakhassane , any suggestions?

File: src/Adaptor.js [hunk 1]
```diff
@@ -197,6 +197,90 @@ export function upsert(table, fields) {
...
+      if (!rows || rows.length === 0) {
+        console.log('No records provided; skipping upsert.');
+        resolve(state);
+      }
+
+      const columns = Object.keys(rows[0]);
+      const columnList = columns.join(', '); 
+      
+      // 1. construct the INSERT statement
+      const insertStmt = "INSERT INTO ".concat(table, " (").concat(columnList, ")");
```
Comment (stuartc) [inline src/Adaptor.js:240]: Should the table name be escaped over here?

File: src/Adaptor.js [hunk 2]
```diff
@@ -197,6 +197,90 @@ export function upsert(table, fields) {
...
+      
+      // 2. construct the subquery
+      const subqueryAlias = 'a';
+      const selectValues = rows.map(function (r) {
+        let selectColumns = columns.map(function (c) {
+          return "".concat(mysql.escape(r[c]), " as ").concat(c);
+        }).join(', ');
+        return "SELECT ".concat(selectColumns);
+      });
+      const subqueryStmt = selectValues.join(' UNION ');
```
Comment (stuartc) [inline src/Adaptor.js:250]: I'm curious about the rational behind building multiple select statements (as input rows) from data and then joining the sets using `UNION`.

The more traditional way would be to construct an known (ordered) list of keys/columns and create sets of tuples like this:

```
INSERT INTO <<table>> (col1, col2)
    VALUES
    ('val1', 24),
    ('val2', 21),
    ('val3', 29)
    AS new
ON DUPLICATE KEY UPDATE
    col1 = new.col1
```
building up individual tuples as selects seems like extra work..

File: src/Adaptor.js [hunk 3]
```diff
@@ -197,6 +197,90 @@ export function upsert(table, fields) {
...
+      const selectStmt = "SELECT * FROM (".concat(subqueryStmt, ") as ").concat(subqueryAlias); 
+      
+      // 3. construct the ON DUPLICATE KEY statement
+      const updateStmt = columns.map(function (c) {
+        return "".concat(c, " = ").concat(subqueryAlias, ".").concat(c);
+      }).join(', ');
+      const onduplicateStmt = "ON DUPLICATE KEY UPDATE ".concat(updateStmt); 
+      
+      // combine 1, 2 & 3
+      const upsertString = "".concat(insertStmt, " ").concat(selectStmt, " ").concat(onduplicateStmt);
```
Comment (stuartc) [inline src/Adaptor.js:260]: I'm concerned why we aren't using the query builder here (and in other places), we're leaving people wide open to sql injection security issues.

The `squel` library used in this adaptor provides all the APIs required to build up this query (and ensure each input is sanitised:

```

var insertParams = squelMysql
    .insert({
      autoQuoteFieldNames: true,
    })
    .into(table)
    .setFields(valuesObj)
    .toParam()
    .onDupUpdate('field1 = field1 + 1')
    .onDupUpdate('field2', 'value2', {dummy: true})
    .onDupUpdate('field3', 'value3')
```

:point_up: this was lifted from the `upsert` function, with stuff I found in the `squel` library...
https://github.com/hiddentao/squel/blob/afa1cb52a61e89404e193bf4e36c2e79dd21c020/test/mysql.test.coffee#L58

File: src/Adaptor.js [hunk 4]
```diff
@@ -197,6 +197,90 @@ export function upsert(table, fields) {
...
+          return "".concat(mysql.escape(r[c]), " as ").concat(c);
+        }).join(', ');
+        return "SELECT ".concat(selectColumns);
+      });
+      const subqueryStmt = selectValues.join(' UNION ');
+      const selectStmt = "SELECT * FROM (".concat(subqueryStmt, ") as ").concat(subqueryAlias); 
+      
+      // 3. construct the ON DUPLICATE KEY statement
+      const updateStmt = columns.map(function (c) {
+        return "".concat(c, " = ").concat(subqueryAlias, ".").concat(c);
```
Comment (stuartc) [inline src/Adaptor.js:255]: Please don't use `.concat` either use arithmetic operators `..." + a + "..` or the preferred ES2015 template string notation. ``` `...${a}...` ```

**Review by stuartc — CHANGES_REQUESTED** [review]
I think a few changes need to be made here.

@taylordowns2000 I can see some of code has been replicated from the `upsert` function, which also uses string interpolation to create the `ON DUPLICATE` clause - we _really_ shouldn't be building strings by hand when talking to a db in this situation. Is there a reason the `onDupUpdate` method wasn't used in the first `upsert`.

At first glance it looks like the `upsert` function is completely capable of handling multiple records (or at least with some very small changes - now that I've seen @dsurrao's work). Would that be preferable.

**Comment by lakhassane** [comment]
Hey @dsurrao thanks for this PR.
Considering the changes suggested by Stu I think we can come up pretty well with that list:
1. drop some of the string manipulation functions in place of string interpolation as it might be a better standard accross all our adaptors.
2. Drop building a set of `select`statements. I think a (easier?) way to build the insert string with multiple values would be to leverage this function from `squel`: `setFieldsRows(array)`
A snippet like this: 
```js
var insertParams = squelMysql
      .insert({
        autoQuoteFieldNames: true,
      })
      .into(table)
      .setFieldsRows(valuesArray)
      .toParam();
```
would generate this sql code dependening on the content of `valuesArray` (`[{name: 'Mamadou', age: 18}, {name: 'Taylor', age: 15}]`)
```
// INSERT INTO ${table} values ('mamadou', 18), ('taylor', 15);
```
3. For the last part which is anything after `ON DUPLICATE KEY UPDATE`, I'm not sure `onDupUpdate()` can do the trick as we don't know beforehand the list of fields to update when duplication arises so might be needed to build that part by hand.

I think it sums everything. Whenever thoses are done, we will merge this and run on tests. Don't hesitate at all for any needed clarifications.

File: src/Adaptor.js (hunk 2 above)
Comment (dsurrao) [reply src/Adaptor.js:250]: The above insert statement is not valid MySQL syntax.

The SELECT..UNION approach builds a query as recommended in MySQL documentation:

INSERT INTO t1
  SELECT * FROM (SELECT c, c+d AS e FROM t2) AS dt
  ON DUPLICATE KEY UPDATE b = e;

File: src/Adaptor.js (hunk 4 above)
Comment (dsurrao) [reply src/Adaptor.js:255]: These concat statements were not in the source code but were generated in the transpiled version in the lib directory.

File: src/Adaptor.js (hunk 3 above)
Comment (dsurrao) [reply src/Adaptor.js:260]: Using the `squel` library with multiple `onDupUpdate` statements is a good approach, I will give this a try.

File: src/Adaptor.js (hunk 1 above)
Comment (dsurrao) [reply src/Adaptor.js:240]: Agree, they should be escaped.

**Comment by dsurrao** [comment]
Hi @stuartc, @lakhassane 
I agree with using a query builder library rather than doing it by hand, but the `squel` library has limitations in this case. Generating code for the `INSERT` statement and building the `ON DUPLICATE UPDATE` line by hand will not work because the use of column aliases is required as described [here](https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html). I cannot find a workaround?

Also, as `squel` is no longer actively maintained, an alternative is to use the [knex](https://knexjs.org/) library, which does support building upsert queries with multiple rows.

This code snippet will generate the output below it:
```
const knex = require('knex')({
    client: 'mysql',
    connection: {
      host : '127.0.0.1',
      user : 'your_database_user',
      password : 'your_database_password',
      database : 'myapp_test'
    }
  });

let query = knex('inserttst')
  .insert([
      { id: 45, fname: "jdoe"},
      { id: 46, fname: "kclay"}
  ])
  .onConflict('id')
  .merge();

console.log(query.toString());
```

Output:
```
insert into `inserttst` (`fname`, `id`) values ('jdoe', 45), ('kclay', 46) 
on duplicate key update `fname` = values(`fname`), `id` = values(`id`)
```

However, even the `fname = values(fname)` syntax is deprecated. 

Only the aliased subquery approach is supported for future versions, but I assume that `knex` will keep up with the changes.

Please let me know your thoughts.

File: src/Adaptor.js (hunk 2 above)
Comment (stuartc) [reply src/Adaptor.js:250]: I can assure you the above is valid MySQL syntax.

https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html

```
CREATE TABLE `names` (
  `id` int NOT NULL AUTO_INCREMENT,
  `first_name` varchar(100) CHARACTER SET utf8mb4 COLLATE utf8mb4_0900_ai_ci DEFAULT NULL,
  `last_name` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`)
);

INSERT INTO `names` (id, first_name, last_name)
    VALUES
    (1, 'val1', 'l1'),
    (2, 'val2', 'l2'),
    (3, 'val3', 'l3')
    AS new;
  
SELECT * FROM names;

id|first_name|last_name|
--+----------+---------+
 1|val1      |l1       |
 2|val2      |l2       |
 3|val3      |l3       |

INSERT INTO `names` (id, first_name, last_name)
    VALUES
    (1, 'foo', 'l1'),
    (2, 'val2', 'bar'),
    (3, 'val3', 'l3')
    AS new
ON DUPLICATE KEY UPDATE
    first_name = new.first_name,
    last_name = new.last_name;
    
SELECT * FROM names;

id|first_name|last_name|
--+----------+---------+
 1|foo       |l1       |
 2|val2      |bar      |
 3|val3      |l3       |
```

The documentation recommends using a derived table when inserting from other tables (with different columns), and using a `UNION` to concatenate those sets. 

The `upsertMany` and `upsert` function do not support inserting data from another table, every value is provided as data. So your sample above doesn't apply to this situation.

File: src/Adaptor.js (hunk 4 above)
Comment (stuartc) [reply src/Adaptor.js:255]: But those `concat`s I'm commenting on are in the `src` directory?

**Comment by lakhassane** [comment]
That could be a good idea @dsurrao. Our next step woud have been to update our other functions to use `knex`. If you could anticipate on that then go for it and we will review when done.

File: src/Adaptor.js (hunk 2 above)
Comment (dsurrao) [reply src/Adaptor.js:250]: This syntax works on MySQL 8, but fails when I try it on MySQL 5.6

File: src/Adaptor.js (hunk 4 above)
Comment (dsurrao) [reply src/Adaptor.js:255]: Right, I will remove those

File: src/Adaptor.js (hunk 2 above)
Comment (dsurrao) [reply src/Adaptor.js:250]: While the documentation refers to inserting data from other tables, I see no reason why you cannot generate the data set using a UNION of `SELECT` statements with aliases.

Comment (stuartc) [reply src/Adaptor.js:250]:
```
INSERT INTO `names` (id, first_name, last_name)
    VALUES
    (1, 'foo', 'l1'),
    (2, 'val2', 'bar'),
    (3, 'val3', 'l3')
ON DUPLICATE KEY UPDATE
    first_name = VALUES(first_name),
    last_name = VALUES(last_name);
```

The MySQL 5.6 compatible version of the above example.

Comment (dsurrao) [reply src/Adaptor.js]: Yes, that was the reason. I will used the VALUES() function, thanks!

Comment (stuartc) [reply src/Adaptor.js]: Ok cool, I get it now. Glad we could iron that out!

I'm sure you know that 5.6 is EOL now. But I'll try my best when moving over to `knex` to see if it's smart enough to use either aliases or `VALUES(...)` depending on the server version; and thereby keeping future versions working with your use case. :+1:

**Comment by dsurrao** [comment]
@lakhassane @stuartc I have updated this PR to use the `squel` library exclusively with the VALUES function.

Please note the deprecation warning message that the server will return:
```
'VALUES function' is deprecated and will be removed in a future release. 
Please use an alias (INSERT INTO ... VALUES (...) AS alias) and replace 
VALUES(col) in the ON DUPLICATE KEY UPDATE clause with alias.col 
instead 1287
```

But this should work for all current versions.

I am avoiding using `knex` for now as that is a bigger code rewrite.

**Comment by lakhassane** [comment]
@dsurrao this looks great. I will merge it and test locally. And yes thanks for flagging the deprecation message.

**Comment by dsurrao** [comment]
Thanks @lakhassane and @stuartc for the helpful feedback! Let me know if I need to do anything else.

**Comment by taylordowns2000** [comment]
@lakhassane , @stuartc , I know that @dsurrao 's team is hoping to use this as part of an OpenFn.org project—could we also cut a release and add this to the drop-down list of recommended adaptors on the platform?

**Comment by stuartc** [comment]
@dsurrao thanks for the work here! I've released `v1.2.0` to [npm](https://www.npmjs.com/package/@openfn/language-mysql) and the package should be available on production shortly (remember to refresh to see it on the adaptor version dropdown) :raised_hands:

**Comment by dsurrao** [comment]
Thanks @stuartc !

**Pull request merged**